haskell/cabal.bzl

```
@@ -916,67 +927,93 @@ def _compute_dependency_graph(repository_ctx, snapshot, core_packages, versioned
-    stack = [stack_cmd, "--resolver", snapshot]
-    if unversioned_packages:
-        _execute_or_fail_loudly(repository_ctx, stack + ["unpack"] + unversioned_packages)
-    exec_result = _execute_or_fail_loudly(repository_ctx, ["ls"])
-    unpacked_sdists = exec_result.stdout.splitlines()
-
-    # Determines path to vendored package's root directory relative to stack.yaml.
-    # Note, this requires that the Cabal file exists in the package root and is
-    # called `<name>.cabal`.
-    vendored_sdists = [
-        truly_relativize(
-            str(repository_ctx.path(label.relative(name + ".cabal")).dirname),
-            relative_to = str(repository_ctx.path("stack.yaml").dirname),
-        )
-        for (name, label) in vendored_packages.items()
-    ]
-    package_flags = {
-        pkg_name: {
-            flag[1:] if flag.startswith("-") else flag: not flag.startswith("-")
-            for flag in flags
-        }
-        for (pkg_name, flags) in repository_ctx.attr.flags.items()
-    }
-    stack_yaml_content = struct(resolver = "none", packages = unpacked_sdists + vendored_sdists, flags = package_flags).to_json()
-    repository_ctx.file("stack.yaml", content = stack_yaml_content, executable = False)
     exec_result = _execute_or_fail_loudly(
         repository_ctx,
-        stack + ["ls", "dependencies", "--global-hints", "--separator=-"],
+        stack + ["ls", "dependencies", "json", "--global-hints", "--external"],
     )
-    transitive_unpacked_sdists = []
-    indirect_unpacked_sdists = []
-    remaining_components = dicts.add(_default_components, user_components)
-    for package in exec_result.stdout.splitlines():
-        name = _chop_version(package)
-        version = _version(package)
+    package_specs = json_parse(exec_result.stdout)
+
+    # Collect package metadata
+    remaining_components = dict(**user_components)
+    for package_spec in package_specs:
+        name = package_spec["name"]
```

Similarly we could throw a high-level message if the name/version fields are missing.

---

Likewise addressed by adding a validation step.